### PR TITLE
Decaffeinate `BucketBar` plugin

### PR DESCRIPTION
**Pull request by lyzadanger**
Part of https://github.com/hypothesis/client/issues/2340

This PR decaffeinates the `BucketBar` plugin and extracts two more utility functions pertaining to calculating anchor buckets. It's a large diff already, so I'm leaving further refactor and jQuery removal for a subsequent step.

Pre-existing tests for the plugin are passing. The extracted utility functions have unit tests.

Here come some comments...

### CODE REVIEW

File: src/annotator/plugin/bucket-bar.js [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+export default class BucketBar extends Delegator {
+  constructor(element, options, annotator) {
+    const defaultOptions = {
+      // gapSize parameter is used by the clustering algorithm
+      // If an annotation is farther then this gapSize from the next bucket
+      // then that annotation will not be merged into the bucket
+      // TODO: This is not currently used; reassess
+      gapSize: 60,
+      html: '<div class="annotator-bucket-bar"></div>',
```
Comment (lyzadanger) [inline src/annotator/plugin/bucket-bar.js:30]: Previously on the prototype

File: src/annotator/plugin/bucket-bar.js [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+      // Selectors for the scrollable elements on the page
+      scrollables: ['body'],
+    };
+
+    const opts = { ...defaultOptions, ...options };
+    super($(opts.html), opts);
+
+    this.options = opts;
+    this.buckets = [];
```
Comment (lyzadanger) [inline src/annotator/plugin/bucket-bar.js:39]: `buckets`, `index` and `tabs` were previously set on the prototype, but I can't see any reason they should be accessible/mutable.

File: src/annotator/plugin/bucket-bar.js [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+      $(element).append(this.element);
+    }
+
+    this.annotator = annotator;
+
+    this.updateFunc = () => this.update();
+
+    $(window).on('resize scroll', this.updateFunc);
+
+    // TODO: Verify binding
```
Comment (lyzadanger) [inline src/annotator/plugin/bucket-bar.js:55]: This is stale. I verified binding was tricksy and for now am wrapping a bound `updateFunc` above.

File: src/annotator/plugin/bucket-bar.js [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+      $(this.options.container).append(this.element);
+    } else {
+      $(element).append(this.element);
+    }
+
+    this.annotator = annotator;
+
+    this.updateFunc = () => this.update();
+
+    $(window).on('resize scroll', this.updateFunc);
```
Comment (lyzadanger) [inline src/annotator/plugin/bucket-bar.js:53]: I note that we are reacting to  `resize` and `scroll` on both `window` and `body`

File: src/annotator/plugin/bucket-bar.js [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+  }
+
+  destroy() {
+    $(window).off('resize scroll', this.updateFunc);
+    this.options.scrollables.forEach(scrollable => {
+      $(scrollable).off('resize scroll', this.updateFunc);
+    });
+  }
+
+  // Update sometime soon
```
Comment (lyzadanger) [inline src/annotator/plugin/bucket-bar.js:68]: Not my comment :)

File: src/annotator/plugin/bucket-bar.js [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+      return updated;
+    });
+  }
+
+  _update() {
+    const { above, below, points } = constructPositionPoints(
+      this.annotator.anchors
+    );
+
+    const bucketInfo = buildBuckets(points);
```
Comment (lyzadanger) [inline src/annotator/plugin/bucket-bar.js:85]: Can't easily destructure here because the name `index` for a property.

File: src/annotator/plugin/test/bucket-bar-test.js [hunk 7]
```diff
@@ -0,0 +1,227 @@
+import $ from 'jquery';
```
Comment (lyzadanger) [inline src/annotator/plugin/test/bucket-bar-test.js:1]: Except for mocking and that kind of thing, these are generally unchanged, just decaffeinated.

File: src/annotator/plugin/test/bucket-bar-test.js [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+        return assert.calledWith(
+          bucketBar.annotator.selectAnnotations,
+          anns,
+          true
+        );
+      })
+    );
+  });
+
+  // Yes this is testing a private method. Yes this is bad practice, but I'd
```
Comment (lyzadanger) [inline src/annotator/plugin/test/bucket-bar-test.js:121]: Not my comment, just FYI

File: src/annotator/util/buckets.js [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+
+/**
+ * Find the closest valid anchor in `anchors` that is offscreen in the direction
+ * indicated.
+ *
+ * @param {Anchor[]} anchors
+ * @param {'up'|'down'} direction
+ * @return {Anchor|null} - The closest anchor or `null` if no valid anchor found
+ */
+export function findClosestOffscreenAnchor(anchors, direction) {
```
Comment (lyzadanger) [inline src/annotator/util/buckets.js:47]: This function was already present on `master`, but this module has been moved and types added here.

File: src/annotator/util/buckets.js [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+
+/**
+ * @typedef {import('../../types/annotator').Anchor} Anchor
+ */
+
+/**
+ * A structured Array representing either the top (`startOrEnd` = 1) or the
+ * bottom (`startOrEnd` = -1) of an anchor's highlight bounding box.
+ *
+ * @typedef {[pixelPosition: number, startOrEnd: (-1 | 1), anchor: Anchor]} PositionPoint
```
Comment (lyzadanger) [inline src/annotator/util/buckets.js:11]: The structure of a `PositionPoint` is not entirely intuitive but I'm leaving as-is for the moment.

File: src/annotator/util/buckets.js [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+      // Add a point for the top of this anchor's highlight box
+      points.push([rect.top, 1, anchor]);
+      // Add a point for the bottom of this anchor's highlight box
+      points.push([rect.bottom, -1, anchor]);
+    }
+  }
+
+  // Sort onscreen points by pixel position, secondarily by position "type"
+  // (top or bottom of higlight box)
+  points.sort((a, b) => {
```
Comment (lyzadanger) [inline src/annotator/util/buckets.js:124]: This was previously a private `_collate` method on `BucketBar`

File: src/annotator/util/buckets.js [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+ * @prop {Anchor[]} below - Anchors that are offscreen below
+ * @prop {PositionPoint[]} points - PositionPoints for on-screen anchor
+ *   highlights. Each highlight box has 2 PositionPoints (one for top edge
+ *   and one for bottom edge).
+ */
+
+/**
+ * @typedef BucketInfo
+ * @prop {Array<Anchor[]>} buckets
+ * @prop {number[]} index - Array of (pixel) positions of each bucket
```
Comment (lyzadanger) [inline src/annotator/util/buckets.js:28]: I plan to rename this property (and/or refactor the structure of this data) as soon as is feasible.

File: src/annotator/util/buckets.js [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+/**
+ * Take a sorted set of `points` representing top and bottom positions of anchor
+ * highlights and group them into a collection of "buckets".
+ *
+ * @param {PositionPoint[]} points
+ * @return {BucketInfo}
+ */
+export function buildBuckets(points) {
+  const buckets = /** @type {Array<Anchor[]>} */ (new Array());
+  const bucketPositions = /** @type {number[]} */ (new Array());
```
Comment (lyzadanger) [inline src/annotator/util/buckets.js:151]: This is my local reference to what will ultimately be the `index` property of the returned `BucketInfo`. It's a little less confusing.

File: src/annotator/util/buckets.js [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+
+  // Anchors that are part of the currently-being-built bucket, and a correspon-
```
Comment (lyzadanger) [inline src/annotator/util/buckets.js:153]: It's really hard to explain the way this logic works; I did my best in the comments herein.

File: src/annotator/util/buckets.js [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+      } else {
+        // Increment the number of times we've seen a start/top edge for this
+        // anchor
+        current.counts[anchorIndex]++;
+      }
+    } else {
+      // positionType = 'end'
+      // This is the bottom/end of an anchor that we should have already seen
+      // a top edge for. Decrement the count, representing that we've found an
+      // end point to balance a previously-seen start point
```
Comment (lyzadanger) [inline src/annotator/util/buckets.js:183]: Some of this is pure inference as to what I believe the original algorithm was attempting to do. I'm confident the answers the migrated variant get are correct; I'm less entirely confident that my explanation is accurate to the original author's intent.

File: src/annotator/util/buckets.js [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+    // merging/removing buckets as we iterate over points.
+    const isFirstOrLastPoint =
+      bucketPositions.length === 0 || index === points.length - 1;
+
+    const isLargeGap =
+      bucketPositions.length &&
+      position - bucketPositions[bucketPositions.length - 1] > BUCKET_GAP_SIZE;
+
+    if (current.anchors.length === 0 || isFirstOrLastPoint || isLargeGap) {
+      // Create a new bucket, because:
```
Comment (lyzadanger) [inline src/annotator/util/buckets.js:212]: I suppose this comment is a bit awkward because the latter two are self-documenting by the reference name...

File: src/annotator/util/buckets.js [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+      // Each bucket gets a corresponding entry in `bucketPositions` for the
+      // pixel position of its eventual indicator in the bucket bar
+      bucketPositions.push(position);
+    } else {
+      // Merge buckets
+      // We will remove 2 (usually) or 1 (if there is only one) bucket
+      // from the buckets collection, and re-add 1 merged bucket (always)
+
+      // Always pop off the last bucket
+      const ultimateBucket = buckets.pop() || new Array();
```
Comment (lyzadanger) [inline src/annotator/util/buckets.js:226]: `|| new Array()` here and below are a TS affordance because of strict null even though there's no way to get to this point without `buckets` having at least one Array of `Anchor`s in it.

File: src/annotator/util/test/buckets-test.js [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+        getBoundingClientRect: fakeGetBoundingClientRect,
+      },
+    });
+  });
+
+  afterEach(() => {
+    $imports.$restore();
+  });
+
+  describe('findClosestOffscreenAnchor', () => {
```
Comment (lyzadanger) [inline src/annotator/util/test/buckets-test.js:41]: Tests for this function were already present (no need for review)

File: src/annotator/util/test/buckets-test.js [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+      assert.isNull(findClosestOffscreenAnchor([{ highlights: [] }], 'down'));
+      assert.isNull(findClosestOffscreenAnchor(fakeAnchors, 'down'));
+    });
+  });
+
+  describe('constructPositionPoints', () => {
+    let fakeAnchors;
+    let stubbedInnerHeight;
+
+    beforeEach(() => {
```
Comment (lyzadanger) [inline src/annotator/util/test/buckets-test.js:120]: This could be DRYed out

File: src/annotator/plugin/test/bucket-bar-test.js [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+  });
+};
+
+const createMouseEvent = function (type, { ctrlKey, metaKey } = {}) {
+  // In a modern browser we could use `new MouseEvent` constructor and pass
+  // `ctrlKey` and `metaKey` via the init object.
+  const event = new Event(type);
+  event.ctrlKey = Boolean(ctrlKey);
+  event.metaKey = Boolean(metaKey);
+  return event;
```
Comment (robertknight) [inline src/annotator/plugin/test/bucket-bar-test.js:22]:
```suggestion
 return new MouseEvent(type, { ctrlKey, metaKey });
```

File: src/annotator/plugin/test/bucket-bar-test.js [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+    });
+
+    it('selects the annotations', () => {
+      // Click on the indicator for the non-empty bucket.
+      const bucketEls = nonEmptyBuckets(bucketBar);
+      assert.equal(bucketEls.length, 1);
+      bucketEls[0].dispatchEvent(createMouseEvent('click'));
+
+      const anns = bucketBar.annotator.anchors.map(anchor => anchor.annotation);
+      return assert.calledWith(
```
Comment (robertknight) [inline src/annotator/plugin/test/bucket-bar-test.js:90]:
```suggestion
      assert.calledWith(
```

This is a common issue when decaffeinating tests. The same applies for other occurrences of `return assert...` in this file.

File: src/annotator/plugin/test/bucket-bar-test.js [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+        bucketBar.annotator.selectAnnotations,
+        anns,
+        false
+      );
+    });
+
+    return [
```
Comment (robertknight) [inline src/annotator/plugin/test/bucket-bar-test.js:97]:
```suggestion
    [
```

File: src/annotator/plugin/test/bucket-bar-test.js [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+        const bucketEls = nonEmptyBuckets(bucketBar);
+        assert.equal(bucketEls.length, 1);
+        bucketEls[0].dispatchEvent(
+          createMouseEvent('click', { ctrlKey, metaKey })
+        );
+
+        const anns = bucketBar.annotator.anchors.map(
+          anchor => anchor.annotation
+        );
+        return assert.calledWith(
```
Comment (robertknight) [inline src/annotator/plugin/test/bucket-bar-test.js:112]:
```suggestion
        assert.calledWith(
```

File: src/annotator/plugin/test/bucket-bar-test.js [hunk 24]
```diff
@@ -0,0 +1,227 @@
...
+      })
+    );
+  });
+
+  // Yes this is testing a private method. Yes this is bad practice, but I'd
+  // rather test this functionality in a private method than not test it at all.
+  //
+  // Note: This could be tested using only the public APIs of the `BucketBar`
+  // class using the approach of the "when a bucket is clicked" tests above.
+  return describe('_buildTabs', function () {
```
Comment (robertknight) [inline src/annotator/plugin/test/bucket-bar-test.js:126]:
```suggestion
  describe('_buildTabs', function () {
```

File: src/annotator/plugin/test/bucket-bar-test.js (hunk 8 above)
Comment (robertknight) [reply src/annotator/plugin/test/bucket-bar-test.js:121]: Noted ;)

File: src/annotator/util/buckets.js [hunk 25]
```diff
@@ -0,0 +1,248 @@
+import { getBoundingClientRect } from '../highlighter';
+
+/**
+ * @typedef {import('../../types/annotator').Anchor} Anchor
+ */
+
+/**
+ * A structured Array representing either the top (`startOrEnd` = 1) or the
```
Comment (robertknight) [inline src/annotator/util/buckets.js:8]: I think "tuple" is a common-enough term in JS circles these days for a structured array that you can use that.

File: src/annotator/util/buckets.js (hunk 12 above)
Comment (robertknight) [reply src/annotator/util/buckets.js:28]: Makes sense. This might also address the comment I had in `bucket-bar.js` about unifying the `buckets` and `index` arrays.

File: src/annotator/util/buckets.js [hunk 26]
```diff
@@ -0,0 +1,248 @@
...
+ * box is onscreen will result in two entries in the `points` Array: one
+ * for the top of the highlight box and one for the bottom
+ *
+ * @param {Anchor[]} anchors
+ * @return {PositionPoints}
+ */
+export function constructPositionPoints(anchors) {
+  const aboveScreenAnchors = new Set();
+  const belowScreenAnchors = new Set();
+  const points = /** @type {PositionPoint[]} */ (new Array());
```
Comment (robertknight) [inline src/annotator/util/buckets.js:101]:
```suggestion
  const points = /** @type {PositionPoint[]} */ ([]);
```

Small stylistic suggestion.

File: src/annotator/util/buckets.js [hunk 27]
```diff
@@ -0,0 +1,248 @@
...
+
+/**
+ * Take a sorted set of `points` representing top and bottom positions of anchor
+ * highlights and group them into a collection of "buckets".
+ *
+ * @param {PositionPoint[]} points
+ * @return {BucketInfo}
+ */
+export function buildBuckets(points) {
+  const buckets = /** @type {Array<Anchor[]>} */ (new Array());
```
Comment (robertknight) [inline src/annotator/util/buckets.js:150]:
```suggestion
  const buckets = /** @type {Array<Anchor[]>} */ ([]);
```

Same with other empty arrays in this file.

File: src/annotator/util/test/buckets-test.js [hunk 28]
```diff
@@ -0,0 +1,275 @@
...
+      assert.deepEqual(positionPoints.points[0], [201, 1, fakeAnchors[2]]);
+      // The bottom position of the first on-screen anchor
+      assert.deepEqual(positionPoints.points[1], [251, -1, fakeAnchors[2]]);
+      // The top position of the second on-screen anchor
+      assert.deepEqual(positionPoints.points[2], [301, 1, fakeAnchors[3]]);
+      // The bottom position of the second on-screen anchor
+      assert.deepEqual(positionPoints.points[3], [351, -1, fakeAnchors[3]]);
+    });
+
+    it('sorts on-screen points based on position type secondarily', () => {
```
Comment (robertknight) [inline src/annotator/util/test/buckets-test.js:178]: I found this test description hard to parse. It looks like you meant, "sorts points by position, then type"?

File: src/annotator/util/buckets.js [hunk 29]
```diff
@@ -0,0 +1,248 @@
...
+    } else {
+      // Add a point for the top of this anchor's highlight box
+      points.push([rect.top, 1, anchor]);
+      // Add a point for the bottom of this anchor's highlight box
+      points.push([rect.bottom, -1, anchor]);
+    }
+  }
+
+  // Sort onscreen points by pixel position, secondarily by position "type"
+  // (top or bottom of higlight box)
```
Comment (robertknight) [inline src/annotator/util/buckets.js:123]: The `sort` callback doesn't do the "sort by position type" part of this sentence. Are you assuming that the points are already sorted by position before this code is called, and therefore `points` preserves this assuming it uses a stable sort?

File: src/annotator/util/test/buckets-test.js [hunk 30]
```diff
@@ -0,0 +1,275 @@
...
+        assert.equal(
+          positionPoints.points[i + fakeAnchors.length][2],
+          fakeAnchors[i]
+        );
+        // This point is a "top" point
+        assert.equal(positionPoints.points[i + fakeAnchors.length][1], 1);
+      }
+    });
+  });
+  describe('buildBuckets', () => {
```
Comment (robertknight) [inline src/annotator/util/test/buckets-test.js:203]: Can you add an empty line above this `describe` block.

**Review by robertknight — CHANGES_REQUESTED** [review]
Overall this looks super. The comments and type docs are super helpful. I left some notes with minor corrections. One of my comments about combining the `buckets` and `index` arrays is something that   I think you may have planned to address in a follow-up PR?

File: src/annotator/plugin/bucket-bar.js (hunk 4 above)
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js:53]: I want to make sure this is correct: I'm going to omit `resize` handling from the `scrollables` but leave it on the `window`. (I'll also update the `destroy` method accordingly)

File: src/annotator/plugin/bucket-bar.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js:85]: Yes, I want to restructure the returned value of `buildBuckets` — next round of changes!

File: src/annotator/plugin/bucket-bar.js [hunk 31]
```diff
@@ -0,0 +1,214 @@
...
+    this.buckets = bucketInfo.buckets;
+    this.index = bucketInfo.index;
+
+    // Scroll up
+    this.buckets.unshift([], above, []);
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js:90]: I KNOW, believe me. "My changes" ended at line 87 and the rest is pretty much direct decaffeination output from here on (with errors fixed).

File: src/annotator/plugin/bucket-bar.js [hunk 32]
```diff
@@ -0,0 +1,214 @@
...
+    this.tabs.slice(this.buckets.length).remove();
+    this.tabs = this.tabs.slice(0, this.buckets.length);
+
+    // Create any new tabs if needed.
+    $.each(this.buckets.slice(this.tabs.length), () => {
+      const div = $('<div/>').appendTo(this.element);
+
+      this.tabs.push(div[0]);
+
+      return (
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js:116]: Good catch.

File: src/annotator/plugin/bucket-bar.js [hunk 33]
```diff
@@ -0,0 +1,214 @@
...
+            }
+          })
+      );
+    });
+
+    this._buildTabs();
+  }
+
+  _buildTabs() {
+    this.tabs.each((d, el) => {
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js:166]: I heartily agree.

File: src/annotator/plugin/bucket-bar.js [hunk 34]
```diff
@@ -0,0 +1,214 @@
...
+  }
+  isLower(i) {
+    return i === this.index.length - 2;
+  }
+}
+
+// Export constants
+BucketBar.BUCKET_SIZE = BUCKET_SIZE;
+BucketBar.BUCKET_NAV_SIZE = BUCKET_NAV_SIZE;
+BucketBar.BUCKET_TOP_THRESHOLD = BUCKET_TOP_THRESHOLD;
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js:214]: There is some other unused code and other code I think might be unused in this module—I'll catch this on the next round.

File: src/annotator/util/buckets.js (hunk 26 above)
Comment (lyzadanger) [reply src/annotator/util/buckets.js:101]: And a good one. I have a reason for why this happened (I'd typically use `[]` when initializing an Array myself), but it's boring.

File: src/annotator/util/buckets.js (hunk 29 above)
Comment (lyzadanger) [reply src/annotator/util/buckets.js:123]: Stale comment from before my migration changes. Will adjust.

Comment (lyzadanger) [reply src/annotator/util/buckets.js:123]: Actually, wait, no. How does it not "sort by position type"?  The wording may be confusing here. If the first entry in the Tuple for the compared `Point`s is equal (i.e. they have the same vertical offset position value), this sort function will continue on to the next `Point` value ("type" of point—top or bottom) and sort by _that_. Thus sorting primarily by position, secondarily by type. How would you suggest this comment is worded to be more clear?

**Comment by lyzadanger** [comment]
@robertknight Feedback integration in fixup commits for now if you want to take a look at changes.

File: src/annotator/plugin/bucket-bar.js (hunk 4 above)
Comment (robertknight) [reply src/annotator/plugin/bucket-bar.js:52]: > I want to make sure this is correct: I'm going to omit resize handling from the scrollables but leave it on the window. 

Yes, exactly.

File: src/annotator/util/buckets.js (hunk 29 above)
Comment (robertknight) [reply src/annotator/util/buckets.js:123]: Oh, right. I forgot that we're dealing with entries that are tuples here. One way to make it clearer would be to make it less generic:

```js
  points.sort(([posA, typeA], [posB, typeB]) => {
    if (posA !== posB) {
      return posA - posB;
    } else {
      return typeA - typeB;
    }
  });
```

This version also has the benefit that it doesn't try to compare `Anchor` objects using the `<` operator if two anchors have the same position and type values. Instead the second comparison will return `0` and so the pre-existing sort order will be preserved.

File: src/annotator/plugin/bucket-bar.js [hunk 35]
```diff
@@ -0,0 +1,214 @@
+import $ from 'jquery';
+import { default as Delegator } from '../delegator';
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js]: This has been addressed

File: src/annotator/plugin/bucket-bar.js [hunk 36]
```diff
@@ -0,0 +1,214 @@
...
+  constructPositionPoints,
+  buildBuckets,
+} from '../util/buckets';
+
+const BUCKET_SIZE = 16; // Regular bucket size
+const BUCKET_NAV_SIZE = BUCKET_SIZE + 6; // Bucket plus arrow (up/down)
+const BUCKET_TOP_THRESHOLD = 115 + BUCKET_NAV_SIZE; // Toolbar
+
+// Scroll to the next closest anchor off screen in the given direction.
+const scrollToClosest = function (anchors, direction) {
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js]: This change was made.

File: src/annotator/plugin/bucket-bar.js [hunk 37]
```diff
@@ -0,0 +1,214 @@
...
+      gapSize: 60,
+      html: '<div class="annotator-bucket-bar"></div>',
+      // Selectors for the scrollable elements on the page
+      scrollables: ['body'],
+    };
+
+    const opts = { ...defaultOptions, ...options };
+    super($(opts.html), opts);
+
+    this.options = opts;
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js]: This line has been removed.

File: src/annotator/plugin/bucket-bar.js [hunk 38]
```diff
@@ -0,0 +1,214 @@
...
+    });
+  }
+
+  // Update sometime soon
+  update() {
+    if (this._updatePending) {
+      return;
+    }
+    this._updatePending = requestAnimationFrame(() => {
+      const updated = this._update();
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js:72]: This diff is outdated. Currently, as if a fixup commit:

```
  update() {
    if (this._updatePending) {
      return;
    }
    this._updatePending = true;
    requestAnimationFrame(() => {
      const updated = this._update();
      this._updatePending = false;
      return updated;
    });
  }
```

File: src/annotator/plugin/bucket-bar.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js:83]: No action to take here in this set of changes

File: src/annotator/plugin/bucket-bar.js (hunk 31 above)
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js:88]: I will address this in a subsequent change.

File: src/annotator/plugin/bucket-bar.js [hunk 39]
```diff
@@ -0,0 +1,214 @@
...
+    this.buckets.push([], below, []);
+    this.index.push(
+      window.innerHeight - BUCKET_NAV_SIZE,
+      window.innerHeight - BUCKET_NAV_SIZE + 1,
+      window.innerHeight
+    );
+
+    // Keep track of tabs to keep element creation to a minimum.
+    if (!this.tabs) {
+      this.tabs = $([]);
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js]: So moved.

File: src/annotator/plugin/bucket-bar.js (hunk 32 above)
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js]: Removed.

File: src/annotator/plugin/bucket-bar.js (hunk 33 above)
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js]: This has been fixed.

File: src/annotator/plugin/bucket-bar.js [hunk 40]
```diff
@@ -0,0 +1,214 @@
...
+
+    this._buildTabs();
+  }
+
+  _buildTabs() {
+    this.tabs.each((d, el) => {
+      let bucketSize;
+      el = $(el);
+      const bucket = this.buckets[d];
+      const bucketLength = bucket?.length;
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js:161]: Removing the optional chaining operator here breaks tests, and given that it's not harmful and there are a bucketload of other tweaks to track, I opted not to pursue it. Do you feel that it is important?

File: src/annotator/plugin/test/bucket-bar-test.js [hunk 41]
```diff
@@ -0,0 +1,227 @@
...
+const createMouseEvent = function (type, { ctrlKey, metaKey } = {}) {
+  // In a modern browser we could use `new MouseEvent` constructor and pass
+  // `ctrlKey` and `metaKey` via the init object.
+  const event = new Event(type);
+  event.ctrlKey = Boolean(ctrlKey);
+  event.metaKey = Boolean(metaKey);
+  return event;
+};
+
+describe('BucketBar', function () {
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/plugin/test/bucket-bar-test.js]: Sorry, I missed the "And the same for other..." part of this comment in my last pass. I'll zap the rest of them.

File: src/annotator/plugin/test/bucket-bar-test.js (hunk 20 above)
Comment (lyzadanger) [reply src/annotator/plugin/test/bucket-bar-test.js]: This is done.

File: src/annotator/plugin/test/bucket-bar-test.js (hunk 21 above)
Comment (lyzadanger) [reply src/annotator/plugin/test/bucket-bar-test.js]: This has been done.

File: src/annotator/plugin/test/bucket-bar-test.js (hunk 22 above)
Comment (lyzadanger) [reply src/annotator/plugin/test/bucket-bar-test.js]: Done.

File: src/annotator/plugin/test/bucket-bar-test.js (hunk 23 above)
Comment (lyzadanger) [reply src/annotator/plugin/test/bucket-bar-test.js]: Done.

File: src/annotator/plugin/test/bucket-bar-test.js (hunk 24 above)
Comment (lyzadanger) [reply src/annotator/plugin/test/bucket-bar-test.js]: Done.

File: src/annotator/util/buckets.js (hunk 25 above)
Comment (lyzadanger) [reply src/annotator/util/buckets.js]: Done.

File: src/annotator/util/buckets.js (hunk 26 above)
Comment (lyzadanger) [reply src/annotator/util/buckets.js]: These have been changed.

File: src/annotator/util/buckets.js (hunk 27 above)
Comment (lyzadanger) [reply src/annotator/util/buckets.js]: Done.

File: src/annotator/util/test/buckets-test.js (hunk 28 above)
Comment (lyzadanger) [reply src/annotator/util/test/buckets-test.js]: Updated test description.

File: src/annotator/util/test/buckets-test.js (hunk 30 above)
Comment (lyzadanger) [reply src/annotator/util/test/buckets-test.js:204]: Yes I can.

File: src/annotator/util/test/buckets-test.js [hunk 42]
```diff
@@ -0,0 +1,276 @@
...
+        return {
+          top: 250,
+          bottom: 250,
+        };
+      });
+      const positionPoints = constructPositionPoints(fakeAnchors);
+      for (let i = 0; i < fakeAnchors.length; i++) {
+        // The bottom position for all of the fake anchors is the same, so
+        // those points will all be at the top of the list
+        assert.equal(positionPoints.points[i][2], fakeAnchors[i]);
```
Comment (robertknight) [inline src/annotator/util/test/buckets-test.js:189]: Reading this again I find it counter-intuitive that "top" points have a type of 1 and "bottom" points have a type of -1 and that the sorting is done such that "top" points appear after "bottom" points. I see this was the case in the previous code though.

**Review by robertknight — APPROVED** [review]
LGTM. We've discussed various improvements for this in Slack but I think we should get this merged and go from there.

File: src/annotator/util/test/buckets-test.js (hunk 42 above)
Comment (lyzadanger) [reply src/annotator/util/test/buckets-test.js:189]: I also think it's technically _fragile_ as the whole thing would break down if it saw a bottom point before a top point for a given anchor. That "shouldn't happen", but it still feels funky.

File: src/annotator/util/test/buckets-test.js [hunk 43]
```diff
@@ -0,0 +1,276 @@
...
+        fakeAnchor(), // top: 501
+      ];
+      stubbedInnerHeight = sinon.stub(window, 'innerHeight').value(410);
+    });
+
+    afterEach(() => {
+      stubbedInnerHeight.restore();
+    });
+
+    it('returns an Array of anchors that are offscreen above', () => {
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/util/test/buckets-test.js:137]: Semantically, I think `array` would be better here, as the concept we're testing for here isn't about it being specifically typed as an `Array` but that it correctly generates the right set of things.... if that makes any kind of sense.
